### ISSUE TITLE
JDB data file iterator may read less item bytes than requested

### ISSUE THREAD

**Issue by artemananiev** [hashgraph/hedera-services#12474]
Current code in `DataFileIterator` (in 0.47 and later versions it's in `DataFileIteratorJDB`):

```java
        final int bytesRead = inputStream.read(dataItemBuffer.array(), offset, bytesToRead - offset);
        if (offset + bytesRead != bytesToRead) {
            throw new EOFException("Was trying to read a data item [" + currentDataItem
                    + "] but ran out of data in the file [" + path + "].");
        }
```

According to `InputStream.read()` JavaDoc, the method may not read the requested number of bytes, it just does its best, but not guaranteed. There are several other places in MerkleDb code when data is read from files, and there we use either `MerkleDbFileUtils.completelyRead()` methods, or `InputStream.readNBytes()`, but not in this place for some reason.

PBJ data file iterator doesn't have this problem. There are no other calls to `InputStream.read()` in MerkleDb.

The easiest option is to use `InputStream.readNBytes()` instead.

**Comment by artemananiev** [comment]
It is too late to fix this problem in 0.47, given that 0.47 is about to be released to mainnet. After a node is upgraded from the old 0.46 state to 0.47, the code in question, `DataFileIterator`, will never be run again.

**Comment by artemananiev** [comment]
I am going to close this ticket, because:

* The bug is in `DataFileIterator` (or `DataFileIteratorJDB`) code, which is only triggered on migration from the old 0.46 state. It will never happen again after mainnet is updated to 0.47
* This code is only used in tests or when a data index needs to be restored on startup, either because the index file on disk is missing, or because it's requested explicitly. Neither should happen in production
* JDB and all related code will be removed from the repo shortly
